File: Python3/프로그래머스/3/42579. 베스트앨범/베스트앨범.py
import collections

def solution(genres, plays):
    ans = []
    
    genres_count = collections.defaultdict(int)
    genres_songs = collections.defaultdict(list)
    
    for idx, genres in enumerate(genres):
        genres_count[genres] += plays[idx]
        genres_songs[genres].append((idx, plays[idx]))
    
    for genre in sorted(genres_count, key= lambda x: -genres_count[x]):
        # print(genre)
        genres_songs[genre].sort(key= lambda x: (-x[1], x[0]))
        
        for genre in genres_songs[genre][:2]:
            ans.append(genre[0])
                
    return ans

File: Swift/프로그래머스/3/42579. 베스트앨범/베스트앨범.swift
import Foundation

func solution(_ genres:[String], _ plays:[Int]) -> [Int] {
    var genre_cnt: [String: Int] = [:]
    var genre_songs: [String: [(Int, Int)]] = [:]
    
    for i in 0..<genres.count {
        let genre = genres[i]
        let play = plays[i]
        
        genre_cnt[genre, default: 0] += play
        genre_songs[genre, default: []].append((i, play))
    }
    
    var ans: [Int] = []
    // print(genre_cnt.sorted { $0.1 > $1.1 })
    for genre in genre_cnt.sorted { $0.1 > $1.1 }.map { $0.0 } {
        // print(genre)
        let songs = genre_songs[genre]!.sorted { $0.0 < $1.0 }.sorted { $0.1 > $1.1 }
        
        for i in 0...1 {
            if songs.count > i {
                ans.append(songs[i].0)
            }
        }
    }
    
    return ans
}